### PR TITLE
Editor support for Sprite scaling & migrate Label scale to ComponentDesc

### PR DESCRIPTION
**Pull request by matgis**
This PR introduces Sprite scaling in the editor, but mostly these are the editor changes required to migrate the `scale` property from  the `LabelDesc` to the `ComponentDesc`. Sprite scaling involved changing a few lines here and there, but as I expected, the scale property migration required *a lot* of changes. I've rewritten tests that were checking the wrong thing, added a mechanism allowing components to modify the `ComponentNode` that hosts them, introduced deep-sanitation of component data embedded in game objects and collections, and added a new implementation of `g/update-property` to the graph internals.

I'm not entirely happy with how it turned out, but I am happy with the way it works. Migrations that span several files (i.e. a `.go` referencing a `.label` file) will be dirty-flagged immediately after the project has loaded to ensure we don't end up with a partial migration, whereas migrations that reside in a single file (embedded label component) will remain unchanged until the user edits something in that file. Scale values will only be written to the `ComponentDesc` when the scale diverges from the default in order to keep diffs minimal.

The migration unfortunately introduces a slight increase to project load times. In my measurements it was around a 11% increase in the large project I tested with. I hope to work on improving this, but I think the feature is shippable despite this.

#### Potential future improvements:
* Get rid of the string decode, decode cycle required to deep-sanitize embedded component data. This is likely where most of the increased load times are coming from.
* Change the `protobuf/read-text` implementation to not return fake default values for message-type fields. I think we want this to have the same semantics as the compiled protobuf classes, which return `nil` in that case.
* Split up loading into finer steps. Something like:
    1. Read the raw protobuf data.
    2. Sanitize it. We'll keep this around to detect unsaved changes.
    3. Apply forced data migrations. These will introduce unsaved changes after load.
    4. Parse the sanitized & migrated data for load dependencies.
    5. Create uninitialized resource nodes in the graph.
    6. Run the `load-fn` on each resource node and its sanitized / migrated data in order.

---

### User-facing changes
* You can now scale Sprite components in the editor.
* File formats: The scale value previously stored in `.label` resources have been migrated to the component entries inside the referencing `.go` or `.collection` files. If your project contains `.go` or `.collection` files that refer to a scaled `.label`, your project will have unsaved changes to the files involved immediately after loading the project in the editor. After saving, the scale value will be stripped from the `.label` and written to each referencing component in all `.go` or `.collection` files, alongside any other data migrations not already committed to these files. You might want to designate a single person on your team to the task of saving and committing the migrated files to source control to avoid merge conflicts.
* Editor users may experience a slight increase in project load times due to the necessary data migration. We intend to address this in a future release.

### Technical changes
* Embedded string data inside collections and game objects is now sanitized on load.
* When registering a component, you can optionally specify a `:alter-referenced-component-fn` to inject some transaction steps after the referenced resource node has been loaded. It will be called with the node ids of the `ReferencedComponent` node and the resource node it refers to (i.e. the `:node-type` you specified when registering the resource type). The function is expected to return a sequence of transaction steps to execute after the referenced node has been loaded.
* When registering a component, you can optionally specify a `:sanitize-embedded-component-fn` to make component-specific changes to the `EmbeddedComponentDesc` and the data for your `:ddf-type` prior to loading, when it is used as an embedded component.

### CODE REVIEW

File: editor/src/clj/editor/workspace.clj [hunk 1]
```diff
@@ -136,6 +139,8 @@ ordinary paths."
                        :dependencies-fn dependencies-fn
                        :write-fn write-fn
                        :read-fn read-fn
+                       :read-raw-fn (or read-raw-fn read-fn)
+                       :sanitize-fn sanitize-fn
```
Comment (matgis) [inline editor/src/clj/editor/workspace.clj:143]: We now expose the `:sanitize-fn` in the `resource-type` map. We use it in a few places in this PR.

File: editor/test/integration/build_test.clj [hunk 2]
```diff
@@ -129,7 +129,7 @@
                 :test-fn (fn [pb targets]
                            (is (= {:color [1.0 1.0 1.0 1.0],
                                    :line-break false,
-                                   :scale [1.0 1.0 1.0 0.0],
+                                   :scale [0.0 0.0 0.0 0.0], ; Default from legacy field added by editor.protobuf/field-desc-default. Not in actual Label$LabelDesc.
```
Comment (matgis) [inline editor/test/integration/build_test.clj:132]: `editor.protobuf/field-desc-default` is the function responsible for emitting fake default values for message-type fields in maps returned by `protobuf/read-text`. As stated above, I think we want to get rid of this behavior in the future.

File: editor/test/integration/collection_test.clj [hunk 3]
```diff
@@ -250,3 +248,144 @@
...
+
+(defn- vector3-pb
+  ^DdfMath$Vector3 [^double x ^double y ^double z]
+  (-> (DdfMath$Vector3/newBuilder)
+      (.setX x)
+      (.setY y)
+      (.setZ z)
+      (.build)))
+
+(deftest component-ddf-scale
```
Comment (matgis) [inline editor/test/integration/collection_test.clj:260]: Because of the "fake defaults" behavior of `editor.protobuf/read-text` (that I'm too scared to change in this already substantial PR), this test is written in terms of the generated protobuf Java classes. This enables us to verify exactly which values get written to the resulting pb-text and binary files.

File: editor/test/integration/game_object_test.clj [hunk 4]
```diff
@@ -60,18 +61,24 @@
...
+                  loaded-embedded-components (:embedded-components load-value)
                   [only-in-saved only-in-loaded] (data/diff saved-embedded-components loaded-embedded-components)]
               (is (nil? only-in-saved))
-              (is (nil? only-in-loaded)))))))))
+              (is (nil? only-in-loaded))
+              (when (or (some? only-in-saved)
+                        (some? only-in-loaded))
+                (println "When comparing" (:label resource-type))
+                (prn 'disk only-in-loaded)
+                (prn 'save only-in-saved)))))))))
```
Comment (matgis) [inline editor/test/integration/game_object_test.clj:84]: This test was testing the wrong thing by being too low-level. Rewritten to make use of the full `read-fn` and `save-data` output of the `GameObjectNode`.

File: editor/test/integration/label_test.clj [hunk 5]
```diff
@@ -90,10 +91,10 @@
                            calls-by-fn)]))
         render-calls-by-pass))
 
-(deftest label-batch-render
+(deftest label-batch-render-test
   (test-util/with-loaded-project
     (let [make-restore-point! #(test-util/make-graph-reverter (project/graph project))
-          add-label-component! (partial test-util/add-embedded-component! app-view (fn [node-ids] (app-view/select app-view node-ids)) (workspace/get-resource-type workspace "label"))
+          add-label-component! (partial test-util/add-embedded-component! app-view (workspace/get-resource-type workspace "label"))
```
Comment (matgis) [inline editor/test/integration/label_test.clj:97]: Got rid of the `select-fn` argument to `test-util/add-embedded-component!`, because no-one was doing anything interesting with it and I wanted `add-embedded-component!` to return the node-id of the created `EmbeddedComponentNode` regardless of what the `select-fn` was doing.

File: editor/test/integration/reload_test.clj [hunk 6]
```diff
@@ -766,9 +766,9 @@
     (let [[workspace project] (setup-scratch world)]
       (let [node-id (project/get-resource-node project "/label/label.label")]
         (is (= "Original" (g/node-value node-id :text)))
-        (is (= [1.0 1.0 1.0] (g/node-value node-id :scale)))
         (is (= [1.0 1.0 1.0 1.0] (g/node-value node-id :color)))
         (is (= [0.0 0.0 0.0 1.0] (g/node-value node-id :outline)))
+        (is (= [0.0 0.0 0.0 1.0] (g/node-value node-id :shadow)))
```
Comment (matgis) [inline editor/test/integration/reload_test.clj:771]: We no longer have `:scale` on the `LabelNode`, but I added a check for `:shadow` so the other checks wouldn't feel lonely after `:scale` left. 🙂

File: editor/test/integration/save_test.clj [hunk 7]
```diff
@@ -35,40 +34,15 @@
...
-                           "collection" GameObject$CollectionDesc
-                           "gui" Gui$SceneDesc
-                           "label" Label$LabelDesc
-                           "model" ModelProto$ModelDesc
-                           "particlefx" Particle$ParticleFX
-                           "tilegrid" Tile$TileGrid
-                           "tilemap" Tile$TileGrid
-                           "tileset" Tile$TileSet
-                           "tilesource" Tile$TileSet})
-
```
Comment (matgis) [inline editor/test/integration/save_test.clj:62]: We can use the `:read-raw-fn` of the `resource-type` map instead of this hard-coded mapping now. 🥳

File: editor/test/integration/save_test.clj [hunk 8]
```diff
@@ -35,40 +34,15 @@
...
 (deftest save-all
   ;; These tests bypass the dirty check in order to verify that the information
   ;; saved is equivalent to the data loaded. Failures might signal that we've
   ;; forgotten to read data from the file, or somehow we're not writing all
-  ;; properties to disk. If you get failures after modifying a .proto file,
-  ;; first make sure your protobuf file extension maps to the corresponding
-  ;; class in the ext->proto map above. This ensures added fields will use
-  ;; defaults from the .proto file and not be reported as missing data in the
-  ;; on-disk reference files.
+  ;; properties to disk.
```
Comment (matgis) [inline editor/test/integration/save_test.clj:45]: Look! I don't have to explain this stupidity anymore! 😄

File: editor/test/integration/save_test.clj [hunk 9]
```diff
@@ -119,21 +93,32 @@
...
+                  (is (nil? only-in-save-pb-msg))
+                  (when (or (some? only-in-disk-pb-msg)
+                            (some? only-in-save-pb-msg))
+                    (println "When comparing" (resource/proj-path resource))
+                    (prn 'disk only-in-disk-pb-msg)
+                    (prn 'save only-in-save-pb-msg)))
+
+                ;; We don't have a read-fn. Compare the strings.
+                (let [disk-string (slurp resource)]
+                  (is (= disk-string save-string)))))))))))
```
Comment (matgis) [inline editor/test/integration/save_test.clj:121]: This is another test that was testing the wrong thing by being too low-level. Rewrote it to use the `read-fn` (and by extension, perform data sanitation) instead of calling `protobuf/read-text` directly. Sanitation is desired, because it means we don't have to continuously update the reference files if we add fields, etc.

File: editor/test/integration/save_test.clj [hunk 10]
```diff
@@ -150,23 +135,46 @@
...
+                    (print-line-diff!)
+
+                    ;; The read data differs. Print a data diff.
+                    (do
+                      (println "When comparing" path)
+                      (prn 'disk only-in-disk-pb-msg)
+                      (prn 'save only-in-save-pb-msg))))
+
+                ;; We don't have a read-fn. Print a line diff.
+                (print-line-diff!)))))))))
```
Comment (matgis) [inline editor/test/integration/save_test.clj:177]: This test should have the same behavior as before, but was rewritten to provide more output about what differs.

File: editor/test/integration/save_test.clj [hunk 11]
```diff
@@ -224,6 +235,9 @@
                ["/collection/all_embedded.collection" delete-first-child!]
                ["/materials/test.material" (set-prop-fn :name "new-name")]
                ["/collection/components/test.label" (set-prop-fn :text "new-text")]
+               ["/label/test.label" (set-prop-fn :text "new-text")]
+               ["/label/embedded_label.go" (set-prop-fn [0] :text "new-text")]
+               ["/label/embedded_label.collection" (set-prop-fn [0 0] :text "new-text")]
```
Comment (matgis) [inline editor/test/integration/save_test.clj:240]: This is in the `save-dirty` test. These verify that the addition of the `scale` field in `GameObject$ComponentDesc` and `GameObject$EmbeddedComponentDesc` does not cause every game object or collection file with embedded components to become dirty immediately after loading the project. It also verifies that the removal of the `scale` field from `Label$LabelDesc` do not cause the same problem.

File: editor/test/integration/save_test.clj [hunk 12]
```diff
@@ -270,8 +284,23 @@
...
+                    (if (some? read-fn)
+                      (let [disk-pb-msg (read-fn resource)
+                            save-pb-msg (with-open [reader (StringReader. save-string)]
+                                          (read-fn reader))
+                            [only-in-disk-pb-msg only-in-save-pb-msg] (data/diff disk-pb-msg save-pb-msg)]
+                        (prn 'disk only-in-disk-pb-msg)
+                        (prn 'save only-in-save-pb-msg))
+                      (let [disk-string (slurp resource)]
+                        (prn 'disk disk-string)
+                        (prn 'save save-string))))))))
```
Comment (matgis) [inline editor/test/integration/save_test.clj:303]: Added helpful output of the differences in case of a failing test case.

File: editor/test/integration/test_util.clj [hunk 13]
```diff
@@ -633,14 +680,24 @@
...
+  ([workspace proj-path pb-map]
+   (assert (integer? workspace))
+   (assert (.startsWith proj-path "/"))
+   (assert (map? pb-map))
+   (let [resource (resource workspace proj-path)
+         resource-type (resource/resource-type resource)
+         write-fn (:write-fn resource-type)
+         content (write-fn pb-map)]
+     (spit resource content)
+     resource)))
```
Comment (matgis) [inline editor/test/integration/test_util.clj:700]: You can now supply a `pb-map` for the file contents.

File: editor/src/clj/dynamo/graph.clj [hunk 14]
```diff
@@ -599,6 +599,13 @@
   (assert node-id)
   (it/update-property node-id p f args))
 
+(defn update-property-ec
+  "Same as update-property, but injects the in-transaction evaluation-context
+  as the first argument to the update-fn."
+  [node-id p f & args]
+  (assert node-id)
+  (it/update-property-ec node-id p f args))
+
```
Comment (matgis) [inline editor/src/clj/dynamo/graph.clj:608]: I have a situation where I need the *in-transaction* value of the `:legacy-scale` since we're doing all of this at load-time. This is an implementation of `g/update-property` that injects the *in-transaction* `evaluation-context` as the first argument to the `update-fn`.

**Review by vlaaad — APPROVED** [review]
LGTM! For me, it was quite hard to follow, so it's good that you added a lot of tests to ensure the behavior is correct 👍

File: editor/test/integration/game_object_test.clj (hunk 4 above)
[the remark this answers is not in the document]
Comment (matgis) [reply editor/test/integration/game_object_test.clj:84]: Yes. I added them so you get some decent output from the test if it fails. 🙂

File: editor/src/clj/editor/collection.clj [hunk 15]
```diff
@@ -844,20 +858,60 @@
...
+
+(defn- sanitize-instance-data [any-instance-desc property-descs-path]
+  ;; GameObject$InstanceDesc, GameObject$EmbeddedInstanceDesc, or GameObject$CollectionInstanceDesc in map format.
+  (-> any-instance-desc
+      (uniform->non-uniform-scale)
+      (game-object/sanitize-property-descs-at-path property-descs-path)))
+
+(defn- sanitize-embedded-game-object-data [embedded-instance-desc resource-type-map]
+  ;; GameObject$EmbeddedInstanceDesc in map format.
+  (let [{:keys [read-fn write-fn]} (resource-type-map "go")]
```
[the remark this answers is not in the document]
Comment (matgis) [reply editor/src/clj/editor/collection.clj:881]: Nah, I'd rather not introduce global state. Doing it the old-fashioned way makes it work well with snapshots, evaluation-contexts, etc.

File: editor/src/clj/editor/resource_node.clj [hunk 16]
```diff
@@ -158,14 +158,16 @@
           ((protobuf/get-fields-fn (protobuf/resource-field-paths ddf-type)) source-value))))
 
 (defn register-ddf-resource-type [workspace & {:keys [ext node-type ddf-type load-fn dependencies-fn sanitize-fn icon view-types tags tag-opts label] :as args}]
-  (let [read-fn (cond->> (partial protobuf/read-text ddf-type)
+  (let [read-raw-fn (partial protobuf/read-text ddf-type)
+        read-fn (cond->> read-raw-fn
```
[the remark this answers is not in the document]
Comment (matgis) [reply editor/src/clj/editor/resource_node.clj:162]: For now it is, although I call it during sanitation as well - a remnant from when I was experimenting with the order of operations. I do think this is really only what the `:read-fn` should do, though. And then expose the `:sanitize-fn` as a separate step so we can obtain unsanitized data for tests and potentially for data migrations (i.e. for when something moved from one resource to another).

File: editor/src/clj/editor/collection.clj (hunk 15 above)
[the remark this answers is not in the document]
Comment (vlaaad) [reply editor/src/clj/editor/collection.clj:881]: It's not a state though, it's a constant, no?

[the remark this answers is not in the document]
Comment (matgis) [reply editor/src/clj/editor/collection.clj:881]: I don't think so? New resource types could be registered at any time. For example, we do it from the Spine plugin when that library is added to the project. Maybe I'm misunderstanding you? 😅

[the remark this answers is not in the document]
Comment (vlaaad) [reply editor/src/clj/editor/collection.clj:881]: But surely no one will register another "go" resource type?

[the remark this answers is not in the document]
Comment (vlaaad) [reply editor/src/clj/editor/collection.clj:881]: Anyway, this code is fine as it is now :)
